File: algorithm/backtracking/macro_filler.cpp
#include <vector>
#include <iostream>
#include <cassert>
#include <set>
#include <unordered_map>
#include <numeric>

using namespace std;

// ratios["chicken breast"] = 4;
// ratios["shrimp"] = 4;
// ratios["beef"] = 2;

// output1["chicken breast"] = 1;
// output1["shrimp"] = 48;
// output1["beef"] = 6;

// total = 55;

double calculateDeviation(unordered_map<string, int>& quantities,
                        unordered_map<string, int>& ratios) {
    double total = 0;
    for (auto i : quantities) total += i.second;

    double deviation = 0.0;
    for (auto i : quantities) {
        double actualRatio = total > 0 ? (double)i.second / total : 0;
        deviation += abs(actualRatio - (double)ratios[i.first] / 10);
    }

    return deviation;
}

void helper(vector<double>& macros, double target,
    vector<int>& curr,
    vector<vector<int>>& res,
    int idx) {

    if (target <= 10) {
        res.push_back(curr);
        return;
    }

    int n = macros.size();
    if (idx >= n) {
        return;
    }

    int multiplier = 1;
    while (multiplier * macros[idx] < target) {
        curr[idx] = multiplier;
        helper(macros, target - multiplier * macros[idx], curr, res, idx + 1);
        curr[idx] = 1;
        multiplier++;
    }
}

// as long as the we are within 10g of target, we count it as a result.
unordered_map<string, int> macroFiller(unordered_map<string, double>& ingredients, 
                                        unordered_map<string, int>& ratios,
                                                double target) {
    int n = ingredients.size();
    unordered_map<int, string> idx_macro;
    vector<double> macros;
    int idx = 0;
    for (auto i : ingredients) {
        idx_macro[idx++] = i.first;
        macros.push_back(i.second);
    }

    vector<int> curr(n, 0);
    vector<vector<int>> res;
    helper(macros, target, curr, res, 0);

    vector<unordered_map<string, int>> outputs;
    for (vector<int> i : res) {
        unordered_map<string, int> m;
        for (int j = 0; j < i.size(); j++) {
            m[idx_macro[j]] = i[j];
        }
        outputs.push_back(m);
    }

    // for (unordered_map<string, int> i : outputs) {
    //     for (auto j : i) {
    //         cout << j.first << ": " << j.second << endl;
    //     }
    //     cout << endl;
    // }

    double minDeviation = std::numeric_limits<double>::max();
    unordered_map<string, int> resMap;
    for (unordered_map<string, int> i : outputs) {
        double d = calculateDeviation(i, ratios);
        // cout << "min deviation: " << minDeviation << endl;
        if (d <= minDeviation) {
            minDeviation = d;
            resMap = i;
            // for (auto j : i) {
            //     cout << j.first << ": " << j.second << endl;
            // }
            // cout << endl;
        }
    }

    return resMap;
}

int main() {
    unordered_map<string, double> ingredients;
    ingredients["chicken breast"] = 2.33; // 10g
    ingredients["shrimp"] = 2.4;
    ingredients["beef"] = 3.07;

    unordered_map<string, int> ratios;
    ratios["chicken breast"] = 4;
    ratios["shrimp"] = 4;
    ratios["beef"] = 2;
    // ingredients["Quinoa"] = 6.4;
    // ingredients["Noodle"] = 7;
    unordered_map<string, int> res = macroFiller(ingredients, ratios, 140);
    // cout << outputs.empty() << endl;

    // unordered_map<string, int> output1;
    // output1["chicken breast"] = 1;
    // output1["shrimp"]= 48;
    // output1["beef"]= 6;

    // double d = calculateDeviation(output1, ratios);
    // cout << "d: " << d << endl;

    for (auto j : res) {
        cout << j.first << ": " << j.second << endl;
    }
    cout << endl;

    return 0;
}

File: algorithm/backtracking/python/macro_filler.py
"""Macro Filler.

Problem: Given a set of ingredients each contributing some number of grams of
macros per unit, a target total macro amount, and a desired macro-ratio between
ingredients, find integer quantities per ingredient whose total macros land
within 10g of the target and whose resulting ingredient ratio deviates least
from the desired ratio.

Approaches:
    1. Backtracking enumeration + deviation scoring.
       - Enumerate all integer-multiplier combinations that get within 10g of
         target via DFS (helper), then pick the combination whose actual
         ingredient ratio minimizes L1 deviation from the desired ratio.
       - Time: O(product of per-ingredient multiplier ranges); Space: O(n * #combos).
"""

from typing import Dict, List


def calculate_deviation(quantities: Dict[str, int], ratios: Dict[str, int]) -> float:
    total = sum(quantities.values())

    deviation = 0.0
    for name, qty in quantities.items():
        actual_ratio = (qty / total) if total > 0 else 0.0
        deviation += abs(actual_ratio - ratios[name] / 10)

    return deviation


def helper(
    macros: List[float],
    target: float,
    curr: List[int],
    res: List[List[int]],
    idx: int,
) -> None:
    if target <= 10:
        res.append(curr[:])
        return

    n = len(macros)
    if idx >= n:
        return

    multiplier = 1
    while multiplier * macros[idx] < target:
        curr[idx] = multiplier
        helper(macros, target - multiplier * macros[idx], curr, res, idx + 1)
        curr[idx] = 1
        multiplier += 1


# as long as we are within 10g of target, we count it as a result.
def macro_filler(
    ingredients: Dict[str, float],
    ratios: Dict[str, int],
    target: float,
) -> Dict[str, int]:
    n = len(ingredients)
    idx_macro: Dict[int, str] = {}
    macros: List[float] = []
    for i, (name, grams) in enumerate(ingredients.items()):
        idx_macro[i] = name
        macros.append(grams)

    curr = [0] * n
    res: List[List[int]] = []
    helper(macros, target, curr, res, 0)

    outputs: List[Dict[str, int]] = []
    for combo in res:
        m: Dict[str, int] = {}
        for j, val in enumerate(combo):
            m[idx_macro[j]] = val
        outputs.append(m)

    min_deviation = float("inf")
    res_map: Dict[str, int] = {}
    for combo_map in outputs:
        d = calculate_deviation(combo_map, ratios)
        if d <= min_deviation:
            min_deviation = d
            res_map = combo_map

    return res_map


if __name__ == "__main__":
    ingredients = {
        "chicken breast": 2.33,  # 10g
        "shrimp": 2.4,
        "beef": 3.07,
    }

    ratios = {
        "chicken breast": 4,
        "shrimp": 4,
        "beef": 2,
    }

    res = macro_filler(ingredients, ratios, 140)

    for name, qty in res.items():
        print(f"{name}: {qty}")
    print()
